testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -0,0 +1,927 @@
+  //   multiple requests/responses (verifying contents received)
+  //   server transport shutdown triggers client shutdown (via GOAWAY)
+  //   queued message InputStreams are closed on stream cancel
+  //     (and maybe exceptions handled)
+
+  @Test
+  public void serverNotListening() {
+    server = null;
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    verify(mockClientTransportListener, timeout(TIMEOUT_MS)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(statusCaptor.capture());
+    assertCodeEquals(Status.UNAVAILABLE, statusCaptor.getValue());
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    verify(mockClientTransportListener, never()).transportReady();
+  }
+
+  @Test
+  public void clientStartStop() throws Exception {
+    server.start(serverListener);
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    client.shutdown();
+    verify(mockClientTransportListener, timeout(TIMEOUT_MS)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(any(Status.class));
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+  }
+
+  @Test
+  public void clientStartAndStopOnceConnected() throws Exception {
+    server.start(serverListener);
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    verify(mockClientTransportListener, timeout(TIMEOUT_MS)).transportReady();
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    client.shutdown();
+    verify(mockClientTransportListener, timeout(TIMEOUT_MS)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(any(Status.class));
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    assertTrue(serverTransportListener.waitForTermination(TIMEOUT_MS, TimeUnit.MILLISECONDS));
+    server.shutdown();
+    assertTrue(serverListener.waitForShutdown(TIMEOUT_MS, TimeUnit.MILLISECONDS));
+    server = null;
+  }
+
+  @Test
+  public void serverAlreadyListening() throws Exception {
+    client = null;
+    server.start(serverListener);
+    Server server2 = newServer();
+    thrown.expect(IOException.class);
+    server2.start(new MockServerListener());
+  }
+
+  @Test
+  public void openStreamPreventsTermination() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    clientStream.start(mockClientStreamListener);
```

Comment:
Today, okhttp and netty are only considering started streams as preventing termination and newStream is just a constructor. That is the same case for Channel and Call; only started Calls prevent shutdown of the Channel.

`InProcessTransport` is a bit of a mix that is probably buggy, it should probably swap to basing off start(); the newStream() check really relates to work done in start(). `DelayedClientTransport` used newStream() since it was convenient.

The difference between the two approaches doesn't really matter too much in my mind because they are indecipherable in ClientCallImpl since we start() the stream very soon after creating it, and we don't have any other synchronization in place in between.

The only real import difference is what happens if you create a _call_ and never start it. For streams it really doesn't seem to matter since it would only tweak the likelihood of winning a race.